src/borg/archive.py

```diff
@@ -719,38 +719,82 @@ def extract_helper(self, item, path, hlm, *, dry_run=False):
+            # First pass: Build fs chunks list
+            fs_chunks = []
+            offset = 0
+            with backup_io("open"):
+                fs_file = open(fs_path, "rb")
+            with fs_file:
+                for chunk in item.chunks:
+                    with backup_io("seek"):
+                        fs_file.seek(offset)
+                    with backup_io("read"):
+                        data = fs_file.read(chunk.size)
+                    if len(data) != chunk.size:
+                        fs_chunks.append(None)
+                    else:
+                        fs_chunks.append(ChunkListEntry(id=self.key.id_hash(data), size=chunk.size))
+                    offset += chunk.size
 
-                    needs_update = True
-                    if len(data_F) == size:
-                        chunkid_F = self.key.id_hash(data_F)
-                        needs_update = chunkid_A != chunkid_F
+            # Compare chunks and collect needed chunk IDs
+            needed_chunks = []
+            for fs_chunk, item_chunk in zip(fs_chunks, item.chunks):
+                if fs_chunk is None or fs_chunk.id != item_chunk.id:
+                    needed_chunks.append(item_chunk)
 
-                    if needs_update:
-                        chunk_data = b"".join(self.pipeline.fetch_many([chunkid_A], ro_type=ROBJ_FILE_STREAM))
-                        fs_file.seek(chunk_offset)
-                        fs_file.write(chunk_data)
+            if not needed_chunks:
+                return True
 
-                    chunk_offset += size
+            # Fetch all needed chunks and iterate through ALL of them
+            chunk_data_iter = self.pipeline.fetch_many(
+                [chunk.id for chunk in needed_chunks], is_preloaded=True, ro_type=ROBJ_FILE_STREAM
+            )
 
-                fs_file.truncate(item.size)
-                return True
+            # Second pass: Update file and consume EVERY chunk from the iterator
+            offset = 0
+            item_chunk_size = 0
+            with backup_io("open"):
+                fs_file = open(fs_path, "rb+")
+            with fs_file:
+                for fs_chunk, item_chunk in zip(fs_chunks, item.chunks):
+                    with backup_io("seek"):
+                        fs_file.seek(offset)
+                    if fs_chunk is not None and fs_chunk.id == item_chunk.id:
+                        offset += item_chunk.size
+                        item_chunk_size += item_chunk.size
+                    else:
+                        chunk_data = next(chunk_data_iter)
+                        if pi:
+                            pi.show(increase=len(chunk_data), info=[remove_surrogates(item.path)])
+                        with backup_io("write"):
+                            if sparse and not chunk_data.strip(b"\0"):
+                                fs_file.seek(len(chunk_data), 1)  # Seek over sparse section
+                                offset += len(chunk_data)
```

Comment:
Guess we can only do it that way when writing to a new file.
But as you are writing to an existing file, you'll need to overwrite existing content with the zeros.